include/heyoka/mascon.hpp

```diff
@@ -0,0 +1,212 @@
+// Copyright 2020 Francesco Biscani (bluescarni@gmail.com), Dario Izzo (dario.izzo@gmail.com)
+//
+// This file is part of the heyoka library.
+//
+// This Source Code Form is subject to the terms of the Mozilla
+// Public License v. 2.0. If a copy of the MPL was not distributed
+// with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#ifndef HEYOKA_MASCON_HPP
+#define HEYOKA_MASCON_HPP
+
+#include <heyoka/detail/igor.hpp>
+#include <heyoka/expression.hpp>
+#include <heyoka/kw.hpp>
+#include <heyoka/math_functions.hpp>
+#include <heyoka/square.hpp>
+#include <vector>
+
+namespace heyoka
+{
+
+// mascon_points -> [N,3] array containing the positions of the masses (units L)
+// mascon_masses -> [N] array containing the position of the masses (units M)
+// pd, qd, rd -> angular velocity of the asteroid in the frame used for the mascon model (units rad/T)
+//
+// GConst kwarg -> Cavendish constant (units L^3/T^2/M)
+// Note, units must be consistent. Choosing L and M is done via the mascon model, T is derived by the value of G. The
+// angular velocity must be consequent (equivalently one can choose the units for w and induce them on the value of G).
+template <typename... KwArgs>
+inline std::vector<std::pair<expression, expression>> make_mascon_system(KwArgs &&... kw_args)
+{
+    // 1 - Check input consistency (TODO)
+    // 2 - We parse the unnamed arguments
+    igor::parser p{kw_args...};
+    if constexpr (p.has_unnamed_arguments()) {
+        static_assert(detail::always_false_v<KwArgs...>,
+                      "The variadic arguments in the construction of the mascon system contain "
+                      "unnamed arguments.");
+    } else {
+        // G constant (defaults to 1).
+        auto Gconst = [&p]() {
+            if constexpr (p.has(kw::Gconst)) {
+                return expression{number{std::forward<decltype(p(kw::Gconst))>(p(kw::Gconst))}};
+            } else {
+                return expression{number{1.}};
+            }
+        }();
+
+        // mascon_points (no default)
+        std::vector<std::vector<number>> mascon_points;
+        if constexpr (p.has(kw::mascon_points)) {
+            for (const auto &point : p(kw::mascon_points)) {
+                mascon_points.emplace_back(std::vector<number>{number{point[0]}, number{point[1]}, number{point[2]}});
+            }
+        } else {
+            throw std::invalid_argument("mascon_points is missing from the kwarg list!");
```

Comment:
Instead of throwing at runtime, it is better here to produce an error at compile time. You can do that via

```c++
static_assert(detail::always_false_v<KwArgs...>, "Error message");
```

Same in the other occurrences below.